corehq/form_processor/backends/sql/processor.py

```diff
@@ -99,6 +100,29 @@ def hard_delete_case_and_forms(cls, domain, case, xforms):
         case.deleted = True
         publish_case_saved(case)
 
+    @classmethod
+    def new_form_from_old(cls, xml, xform, value_responses_map, user_id):
+        from corehq.form_processor.interfaces.processor import FormProcessorInterface
+        from corehq.form_processor.interfaces.dbaccessors import FormAccessors
+        interface = FormProcessorInterface(xform.domain)
+        existing_form = FormAccessors(xform.domain).get_with_attachments(xform.get_id)
+
+        new_xml = etree.tostring(xml)
+        form_json = convert_xform_to_json(new_xml)
+        new_form = interface.new_xform(form_json)
+        new_form.user_id = user_id
+        new_form.domain = existing_form.domain
+        new_form.app_id = existing_form.app_id
+        new_form.xmlns = existing_form.xmlns
+        new_form.received_on = existing_form.received_on
+        new_form.edited_on = datetime.datetime.utcnow()
+
+        existing_form, new_form = cls.apply_deprecation(existing_form, new_form)
```

Comment:
if you used `corehq.form_processor.parsers.form.apply_deprecation` instead you could get rid of a few more lines:

```diff
-  new_form.received_on = existing_form.received_on
-  new_form.edited_on = datetime.datetime.utcnow()
-   existing_form, new_form = cls.apply_deprecation(existing_form, new_form)
+   existing_form, new_form = apply_deprecation(existing_form, new_form)
-   existing_form = cls.assign_new_id(existing_form)
-   new_form.deprecated_form_id = existing_form.get_id
-   new_form.edited_on = datetime.datetime.utcnow()
```